spec.html

```diff
@@ -39015,57 +39030,11 @@ <h1>Promise.prototype.finally ( _onFinally_ )</h1>
-            1. Let _catchFinally_ be CreateBuiltinFunction(_stepsCatchFinally_, &laquo; [[Constructor]], [[OnFinally]] &raquo;).
-            1. Set _catchFinally_.[[Constructor]] to _C_.
-            1. Set _catchFinally_.[[OnFinally]] to _onFinally_.
-          1. Return ? Invoke(_promise_, `"then"`, &laquo; _thenFinally_, _catchFinally_ &raquo;).
-        </emu-alg>
-
-        <emu-clause id="sec-thenfinallyfunctions">
-          <h1>Then Finally Functions</h1>
-          <p>A Then Finally function is an anonymous built-in function that has a [[Constructor]] and an [[OnFinally]] internal slot. The value of the [[Constructor]] internal slot is a `Promise`-like constructor function object, and the value of the [[OnFinally]] internal slot is a function object.</p>
-          <p>When a Then Finally function _F_ is called with argument _value_, the following steps are taken:</p>
-          <emu-alg>
-            1. Let _onFinally_ be _F_.[[OnFinally]].
-            1. Assert: IsCallable(_onFinally_) is *true*.
-            1. Let _result_ be ? Call(_onFinally_, *undefined*).
-            1. Let _C_ be _F_.[[Constructor]].
-            1. Assert: IsConstructor(_C_) is *true*.
-            1. Let _promise_ be ? PromiseResolve(_C_, _result_).
-            1. Let _valueThunk_ be equivalent to a function that returns _value_.
-            1. Return ? Invoke(_promise_, `"then"`, &laquo; _valueThunk_ &raquo;).
-          </emu-alg>
-          <p>The `length` property of a Then Finally function is *1*.</p>
-        </emu-clause>
-
-        <emu-clause id="sec-catchfinallyfunctions">
-          <h1>Catch Finally Functions</h1>
-          <p>A Catch Finally function is an anonymous built-in function that has a [[Constructor]] and an [[OnFinally]] internal slot. The value of the [[Constructor]] internal slot is a `Promise`-like constructor function object, and the value of the [[OnFinally]] internal slot is a function object.</p>
-          <p>When a Catch Finally function _F_ is called with argument _reason_, the following steps are taken:</p>
-          <emu-alg>
-            1. Let _onFinally_ be _F_.[[OnFinally]].
-            1. Assert: IsCallable(_onFinally_) is *true*.
-            1. Let _result_ be ? Call(_onFinally_, *undefined*).
-            1. Let _C_ be _F_.[[Constructor]].
-            1. Assert: IsConstructor(_C_) is *true*.
-            1. Let _promise_ be ? PromiseResolve(_C_, _result_).
-            1. Let _thrower_ be equivalent to a function that throws _reason_.
-            1. Return ? Invoke(_promise_, `"then"`, &laquo; _thrower_ &raquo;).
-          </emu-alg>
-          <p>The `length` property of a Catch Finally function is *1*.</p>
-        </emu-clause>
+          1. Let _resultCapability_ be ? NewPromiseCapability(_C_, _promise_).
+          1. Return PerformPromiseThen(_promise_, _onFinally_, _onFinally_, _resultCapability_).
```

Comment:
The fact that this does not observably call `.then` immediately means that a subclass has to override `.finally` - with the current design, it only has to override `.then` (which is a requirement)

Reply:
I think that makes sense because in essence `then` and `finally` are different. Could you please clarify why you think otherwise? The current design is based on what I consider to be a workaround instead of being an exact expression of what `finally` means conceptually, hence the extra ticks that `finally` doesn't conceptually require.
I think `finally` should call `then` only if the behavior expressed by `then` is part of `finally`'s behavior. In my opinion it's not fully; they are similar only to a certain degree. Perhaps listing the pros and cons and the possible trade-offs and analyzing which one is better would help making the right choice.

Reply:
To clarify, you mean that if the subclass overrides `then`, it must then override `finally` as well, is that correct? Because a subclass doesn't have to override `then` unless it chooses to.

Reply:
To be fair, no one likes promise subclassing anyway :D

Reply:
Though I am sympathetic to the goal of reducing the number of promises created by `Promise.prototype.finally` in general, I believe it is desirable (and widely expected) that a `Promise` subclass only has to override `then`, and all the other methods will automatically work in terms of the overridden `then`.

Regardless of any other observable semantic differences introduced by that PR, abandoning that property of `Promise` subclassing may be a deal-breaker for any approach that implements `finally` without calling `then`.

Reply:
@benjamn That would imply that whatever method is added to `Promise.prototype` in the future must be expressed in terms of `Promise.prototype.then` which is a limitation whose ground I'm not familiar with. In general a class doesn't have a 'heliocentric' method, and a subclass can override whatever behavior it wants to. I don't expect many people to subclass `Promise` and I certainly expect that people who subclass `Promise` know very well what they're doing, so that having to extend more than one method (if they need to) wouldn't be a deal-breaker.

Reply:
It’s definitely a dealbreaker; i used to have it this way and that was a blocker for proceeding to stage 3.

Reply:
@ljharb I assume there are meeting notes about this. Do you remember during which meeting this was discussed?

Reply:
https://github.com/rwaldron/tc39-notes/blob/b8da60318b564f136cbe8385f17f42abc0666cdd/es7/2017-01/jan-25.md#13iif-promiseprototypefinally-to-stage-3

Reply:
@ljharb Thank you for the link! Reading the notes I understood something else and not that it was a blocker for proceeding to the next stage. DE expressed some doubt about modifying `catch`. DE also said that if `then` is the builtin `then` then some optimization can be done. But if I understand correctly he meant that even though `then` must be looked up, optimizations are still possible. I don't see how not having to look up `then` would result in less optimized runtime. Then yourself said that it's good for `finally` to be consistent with `catch`. JLZ said that Angular might depend on the existing behavior of `catch`. AK didn't want the proposal to proceed because it would have been possible that the spec would be changed after that (he didn't say that he disagreed with `finally` not calling `then`). AWB expressed indifference. MM was being poetical.

People were pretty clear that changing `catch` could be problematic. But what I consider to be the essence of the discussion were having here, namely whether or not `finally` is conceptually expressed in terms of `then`, was not touched in that meeting. Perhaps if this matter was brought into the discussion more clearly, people would have looked at it differently. These notes don't show that this aspect was discussed seriously.

Nobody seemed firm about `finally` having to be consistent with `catch`. Somebody (we don't know who) even asked what was the motivation about not calling `then`. I think yourself were the most 'firm', by saying that it seems that consistency is the way to go, which doesn't seem very firm either. I think consistency with `catch` received a lot more attention than conceptual consistency. And in general the debate gives the impression of uncertainty. This contrasts a lot with saying that 'it was a blocker'.

Reply:
The general consensus in the room - or at least, the only way to achieve consensus, which is effectively the same - was that either a) `.catch` and `.finally` should both directly delegate to `.then`, or, b) `.catch` and `.finally` should both call abstract operations directly. "b" is not web-compatible.

I'm not going to enumerate which committee members were blockers there, both to not throw folks under the bus and because it's irrelevant (the notes are not transcripts, intentionally) - I hope you can simply trust that what I've said above is the case.

Reply:
Can we phrase it so that engines have to implement `finally` through `then` only for non-subclasses? That would be sufficient - subclassing is almost never used anyway since `async` functions return the base class always anyway.

Reply:
Engines can already optimize out the `.then` calls when it's not a subclass and when `Promise.prototype.then` is unmodified; but no, if I replace `Promise.prototype.then` it needs to be able to observe the calls too.

Reply:
@ljharb I don't doubt what you're saying but I also can't help notice that the motivation for why those 2 options were the only ones considered by TC39 is missing. If TC39 isn't willing to evaluate (or re-evaluate, in case they already have) the aspects I've mentioned, this PR is DOA.

Reply:
The third option is what this PR does, which is what the finally proposal did originally, and what the committee rejected - see https://github.com/tc39/proposal-promise-finally/pull/34